src/box/box.cc

```diff
@@ -5030,6 +5031,58 @@ box_read_ffi_enable(void)
 		box_read_ffi_is_disabled = false;
 }
 
+int
+box_generate_space_id(uint64_t *new_space_id)
```

Comment:
Overall, the function code looks confusing. I'd expect `space_cache_find_next_unused_id(id)` to return an id *strictly greater* than the given one (because of *next* in its name) while in fact it returns a value >= id. I assume this is because we use `boxk` to update and get the max id. I think we could rewrite the code like this to make it look more straightforward:

```C
struct tuple *tuple;
if (box_index_get(BOX_SPACE_MAX, 0, key, key_end, &tuple) != 0)
	return -1;
uint32_t max_id = 0;
if (tuple != NULL && tuple_field_u32(tuple, 1, &max_id) != 0)
	return -1;
if (max_id > BOX_SPACE_MAX || max_id < BOX_SYSTEM_ID_MAX)
	max_id = BOX_SYSTEM_ID_MAX;
*new_space_id = space_cache_find_next_next_unused_id(max_id);
if (*new_space_id > BOX_SPACE_MAX) {
	*new_space_id = space_cache_find_next_next_unused_id(
					BOX_SYSTEM_ID_MAX);
	if (*new_space_id > BOX_SPACE_MAX)
		panic("...");
}
if (boxk(IPROTO_REPLACE, BOX_SCHEMA_ID, NULL,
	 "[%s%lu]", "max_id", (unsigned long)*new_space_id) != 0)
	return -1;
return 0;
```

To avoid permission errors, we could change credentials to admin's for `box_index_get()`, like we do here:

https://github.com/tarantool/tarantool/blob/56a43147444141ada97d5f03faa0dd218dbbcc89/src/box/box.cc#L3404-L3410

Reply:
Reworked the function in a more straightforward manner.

Reply:
Also, updated changeling and commit message - I considered it necessary to note the weakening of access requirements.